### PR TITLE
Docker

### PR DESCRIPTION
**Pull request by kevinazhu**
1. Adding build flags to statically link libraries
2. Implemented docker deployment
3. Added make targets to automate docker deployment

NOTE: private key must be placed in home (~) directory for the deployment commands to function properly.
AWS server is currently located at ec2-54-218-71-128.us-west-2.compute.amazonaws.com

### CODE REVIEW

**Comment by dstephan1** [comment]
Do we always want to compile statically?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+RUN apt-get install -y libboost-all-dev
+RUN apt-get install -y make
+RUN apt-get install -y g++
+
+WORKDIR /opt/webserver
+COPY . /opt/webserver
+
+RUN make clean && make
+
+CMD tar -cf - webserver
```
Comment (dstephan1) [inline Dockerfile]: Isn't this the Dockerfile for the large docker image (it has to include dynamic libs and stuff)? Is there one with busybox?

File: Makefile [hunk 2]
```diff
@@ -1,26 +1,27 @@
...
 
 default: webserver
 
 %.o: %.cc $(DEPS)
-	$(CXX) $(CXXFLAGS) $(OPTIMIZE) $(COV) -c -o $@ $< $(BOOSTFLAG)
+	$(CXX) $(CXXFLAGS) $(OPTIMIZE) $(COV) -c -o $@ $<
 
 webserver: main.cc $(OBJ)
-	$(CXX) $(CXXFLAGS) $(OPTIMIZE) $(COV) -o $@ $^ $(BOOSTFLAG)
+	$(CXX) $(CXXFLAGS) $(OPTIMIZE) $(COV) -o $@ $^ $(LDLIBS)
```
Comment (dstephan1) [inline Makefile]: Do we always want to compile with static libraries? It isn't hard to make it optional, but will we ever actually want the dynamic library version?

Comment (kevinazhu) [inline Makefile]: Not really sure tbh... Spec seemed to indicate to just change it, but no idea if that was just vague.

File: Dockerfile (hunk 1 above)
Comment (kevinazhu) [inline Dockerfile]: busybox version gets created within the deploy folder by line 59 in the Makefile

File: reverse_proxy_handler.cc [hunk 3]
```diff
@@ -14,6 +14,7 @@
 RequestHandler::Status ReverseProxyHandler::Init(const std::string& uri_prefix, const NginxConfig& config)
 {
 	prefix_ = uri_prefix;
+	protocol_ = "80";
```
Comment (dstephan1) [inline reverse_proxy_handler.cc]: What is protocol 80? I've never heard of 80://google.com

Comment (kevinazhu) [inline reverse_proxy_handler.cc]: I guess I should refactor it to be port, cuz that's what it actually is for. Protocol being set to http fails when trying to resolve the connection which was what was causing the 500 internal server error before.

File: Dockerfile (hunk 1 above)
Comment (dstephan1) [inline Dockerfile]: Right. It might be more clear in a different file. It'll also make it easier to add lines. This is works fine though.

**Comment by dstephan1** [comment]
If you refactor protocol_ it should be fine to merge. 
We can fix the connection leak later.

**Comment by jteo1** [comment]
good

**Pull request merged**